internal/pkg/rule/function/provider.go

```diff
@@ -0,0 +1,95 @@
+package function
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/google/cel-go/common/types/ref"
+)
+
+var ErrMaxKeys = errors.New("max keys reached")
+
+type StatefulFunction interface {
+	ref.Type
+	ref.Val
+	Call(value any) bool
+}
+
+type state struct {
+	state        any
+	lastAccessed time.Time
+}
+
+type StatefulFunctionProvider struct {
+	StateName               string
+	ctx                     context.Context
+	ctxCancel               context.CancelFunc
+	maxKeys                 int32
+	statefulFunctionBuilder func(state any) StatefulFunction
+	stateBuilder            func() any
+	globalStatefulFunction  StatefulFunction
+	keyedStates             map[string]state
+}
+
+func NewStatefulFunctionProvider(stateName string, statefulFunctionBuilder func(state any) StatefulFunction, stateBuilder func() any) *StatefulFunctionProvider {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &StatefulFunctionProvider{
+		StateName:               stateName,
+		ctx:                     ctx,
+		ctxCancel:               cancel,
+		statefulFunctionBuilder: statefulFunctionBuilder,
+		stateBuilder:            stateBuilder,
+		keyedStates:             make(map[string]state),
+	}
+}
+
+func (sfp *StatefulFunctionProvider) WithManagedKeyedState(ttl time.Duration, maxKeys int32) {
+	sfp.maxKeys = maxKeys
+
+	ticker := time.NewTicker(time.Minute)
+	go func() {
+		for {
+			select {
+			case <-sfp.ctx.Done():
+				return
+			case <-ticker.C:
+				for key, state := range sfp.keyedStates {
+					if time.Since(state.lastAccessed) > ttl {
+						delete(sfp.keyedStates, key)
```

Comment:
is this safe? maps in golang can be read/write concurrently if I recall correctly?

Reply:
I checked it when I implemented that part. It seemed to be valid: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

Reply:
We talked about that. There were multiple writes in different goroutines that could be problematic. A mutex was added to avoid race conditions